File: backend/backend/legacy/cs_chess_algorithms (pre-python-refactorization)/piece_movement_validation.cs
public bool DoesFollowPawnMovementRules(moveInfo) 
{
    //Is path clear
    if(!IsPathClear(moveInfo))
        return false;

    if(moveInfo.distanceX == 0) //Forward move
    {
        if(moveInfo.whiteTurn) {
            //One square and two square move as white
            if(moveInfo.distanceY == 1 || moveInfo.distanceY == 2 && moveInfo.sourceY == 1)
                return true;
        }
        else {
            //One square and two square move as black
            if(moveInfo.distanceY == -1 || moveInfo.distanceY == -2 && moveInfo.sourceY == 6)
                return true;
        }  
    }
    else //Diagonal takes move
    {
        //Standard takes
        if(Math.Abs(moveInfo.distanceX) == 1 && moveInfo.whiteTurn ? moveInfo.distanceY == 1 : moveInfo.distanceY == -1)
            return true;

        //En passant
        if(moveInfo.targetPiece == null && moveInfo.enPassantTarget == ((int?)moveInfo.targetX, (int?)moveInfo.targetY))
            return true;
    }    

    return false;
}

public bool DoesFollowKnightMovementRules(moveInfo) 
{
    //Is target square occupied by friendly piece
    if(moveInfo.targetPiece != null && moveInfo.sourcePiece.isWhite == moveInfo.targetPiece.isWhite)
        return false;

    //Standard movement
    if(((Math.Abs(moveInfo.distanceX) == 2 && Math.Abs(moveInfo.distanceY) == 1) || (Math.Abs(moveInfo.distanceX) == 1 && Math.Abs(moveInfo.distanceY) == 2)))
        return true;

    return false;
}

public bool DoesFollowRookMovementRules(moveInfo) 
{
    //Is path clear
    if(!IsPathClear(moveInfo))
        return false;

    //Standard movement  
    if(moveInfo.distanceX == 0 || moveInfo.distanceY == 0)
        return true;

    return false;
}

public bool DoesFollowBishopMovementRules(moveInfo) 
{
    //Is path clear
    if(!IsPathClear(moveInfo))
        return false;
    
    //Standard movement  
    if(Math.Abs(moveInfo.distanceX) == Math.Abs(moveInfo.distanceY))
        return true;

    return false;
}

public bool DoesFollowQueenMovementRules(moveInfo) 
{
    //Standard movement
    if(DoesFollowRookMovementRules(moveInfo) || DoesFollowBishopMovementRules(moveInfo))
        return true;

    return false;
}

public bool DoesFollowKingMovementRules(moveInfo) 
{
    //Is path clear
    if(!IsPathClear(moveInfo))
        return false;

    //Castling
    if(Math.Abs(moveInfo.distanceX) == 2 && moveInfo.distanceY == 0 && IsCastlingValid(moveInfo))
        return true;

    //Standard movement
    if(Math.Max(Math.Abs(moveInfo.distanceX), Math.Abs(moveInfo.distanceY)) <= 1)
        return true;

    return false;
}

File: backend/backend/python_chess_algorithms/piece_movement_validation.py
from .data_structures import ChessPiece, ChessFigure, MoveInfo

def is_path_clear(move_info):
    if move_info.target_piece is not None:
        # Checking pawn forward movement collision
        if move_info.source_piece.type == ChessPiece.Pawn and move_info.direction_x == 0:
            return False
        # Checking for a friendly piece on the final square
        elif move_info.target_piece.is_white == move_info.source_piece.is_white:
            return False
    
    # For knight the validation ends here, we don't check in-between squares occupancy due to the knight "jumping" over other pieces
    if move_info.source_piece.type == ChessPiece.Knight:
        return True

    # Checking all in-between squares occupancy
    for i in range(1, max(abs(move_info.distance_x), abs(move_info.distance_y))):
        if move_info.chess_board[move_info.source_x + i * move_info.direction_x][move_info.source_y + i * move_info.direction_y] is not None:
            return False
        
    return True


def does_follow_pawn_movement_rules(move_info):
    if move_info.direction_x == 0:
         # One or two steps forward for white pawns
        if move_info.source_piece.is_white:
            if move_info.distance_y == -1 or (move_info.distance_y == -2 and move_info.source_y == 6):
                return is_path_clear(move_info)
        # One or two steps forward for black pawns
        elif move_info.distance_y == 1 or (move_info.distance_y == 2 and move_info.source_y == 1):
            return is_path_clear(move_info)        
    elif abs(move_info.distance_x) == 1 and abs(move_info.distance_y) == 1:
        # Standard capture
        if move_info.target_piece is not None:
            if (move_info.source_piece.is_white and move_info.distance_y == -1 or not move_info.source_piece.is_white and move_info.distance_y == 1):
                return is_path_clear(move_info)
        # En passant
        elif move_info.en_passant_target == (move_info.target_x, move_info.target_y):
            return True
                
    return False


def does_follow_knight_movement_rules(move_info):
    # Standard knight movement
    if (abs(move_info.distance_x) == 2 and abs(move_info.distance_y) == 1) or (abs(move_info.distance_x) == 1 and abs(move_info.distance_y) == 2):
        return is_path_clear(move_info)

    return False


def does_follow_rook_movement_rules(move_info):
    # Standard rook movement (along straight lines)
    if move_info.distance_x == 0 or move_info.distance_y == 0:
        return is_path_clear(move_info)

    return False


def does_follow_bishop_movement_rules(move_info):
    # Standard bishop movement (along diagonals)
    if abs(move_info.distance_x) == abs(move_info.distance_y):
        return is_path_clear(move_info)

    return False


def does_follow_queen_movement_rules(move_info):
    # Queen moves like a rook or bishop
    if does_follow_rook_movement_rules(move_info) or does_follow_bishop_movement_rules(move_info):
        return True

    return False


def does_follow_king_movement_rules(move_info):
    # Castling
    from .move_validation_algorithms import is_castling_valid
    if abs(move_info.distance_x) == 2 and move_info.distance_y == 0 and is_castling_valid(move_info):
        return is_path_clear(move_info)

    # Standard king movement (to adjacent square)
    if max(abs(move_info.distance_x), abs(move_info.distance_y)) <= 1:
        return is_path_clear(move_info)

    return False